### PR TITLE
fix: #1269: Extra logging in matching algorithm

### PR DESCRIPTION
**Pull request by SergeyEzh**

### CODE REVIEW

**Review by zabeen — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 7 unresolved discussions (waiting on @SergeyEzh)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_R4FCCrC9o0ykUKHg):*
nitpick: should be `chore`



_Code quote:_
```
fix
```
___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 85 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_RgMFBjPxURo50xdh:-Nu_RgMFBjPxURo50xdi:b-76twet) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L85)):*
> ```Smalltalk
>             // to do so would be to remove ranking of results, and may cause issues down the line where all results *do* need to be loaded into memory.
>             var scoredMatches = await scoringService.StreamScoring(request);
>             searchLogger.SendTrace($"{nameof(IMatchScoringService.StreamScoring)} has prepared enumeration to consume", LogLevel.Verbose);
> ```

"has completed scoring"



_Code quote:_
```Smalltalk
has prepared enumeration to consume
```
___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 86 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_RHAz0_pfZSRBCPXJ:-Nu_RHAz0_pfZSRBCPXK:b-fqc5jg) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L86)):*
> ```Smalltalk
>             var scoredMatches = await scoringService.StreamScoring(request);
>             searchLogger.SendTrace($"{nameof(IMatchScoringService.StreamScoring)} has prepared enumeration to consume", LogLevel.Verbose);
>             var reifiedScoredMatches = scoredMatches.DistinctBy(m => m.MatchResult.DonorId).ToList(); // Should we load all 300k results in memory despite all batching?
> ```

I agree this doesn't look great, and I hope it is the cause of our issues now we've spotted it!

Add a TODO comment to this new ticket, we can fix it after our initial investigation:
https://github.com/Anthony-Nolan/Atlas/issues/1271



_Code quote:_
```Smalltalk
// Should we load all 300k results in memory despite all batching?
```
___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 93 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_TXpC6P5WKJWzZJLD:-Nu_TXpD-3OC5ycUrl6o:bslnol7) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L93)):*
> ```Smalltalk
>             searchLogger.SendTrace("Donor lookup has been prepared", LogLevel.Verbose);
> 
>             var resultsFilteredByDonorDetails = donorDetailsResultFilterer.FilterResultsByDonorData(        // In order to speed up searches and reduce memory used this filtering should be applied before stage 1&2. 
> ```

what do you mean by before stage 1? 

I think filtering by registry code is being done this late, because we didn't use to store registry code in the matching db until a few months ago, and had to fetch the data from the main donor db - so we did that as a one-off step to avoid repeated calls to the external db.

Now that we have that data "locally" we could shift the filtering to same place all the other filtering is done:
https://github.com/Anthony-Nolan/Atlas/blob/master/Atlas.MatchingAlgorithm/Services/Search/Matching/MatchingService.cs#L104

I'm not sure it would improve performance, but it would be cleaner.

Again, think we should do that in #1271



_Code quote:_
```Smalltalk
filtering should be applied before stage 1&2
```
___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 98 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu__Alw-r11A2U1Ta9u:-Nu__Alw-r11A2U1Ta9v:bcjsxqz) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L98)):*
> ```Smalltalk
>                 donorLookup
>             ).ToList();
>             searchLogger.SendTrace($"{nameof(IDonorDetailsResultFilterer.FilterResultsByDonorData)} has completed building results list", LogLevel.Verbose);
> ```

filtering results



_Code quote:_
```Smalltalk
building results list
```
___
*[`Atlas.MatchingAlgorithm/Services/Search/Matching/DonorMatchingService.cs` line 114 at r1](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-NuZPPDJ7pDswYRnrpxG:-NuZPPDJ7pDswYRnrpxH:bpcy766) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/ce5bad879a712b56956b6d85637abc3259188b0d/Atlas.MatchingAlgorithm/Services/Search/Matching/DonorMatchingService.cs#L114)):*
> ```Smalltalk
>                         // Loads all results from previous loci into memory. This is expected to be very memory intensive, and is only necessary for searches that allow two mismatches at all required loci, e.g. 8/10, 4/8
>                         // If this still causes OutOfMemory exceptions, we may need to look into Zipping multiple ordered streams from different loci.
>                         : (await System.Linq.AsyncEnumerable.ToListAsync(previousLociResultStream)).ToDictionary(x => x.DonorId, x => x);
> ```

suggest tidying up by moving this line, along with the two new logging steps, to a new method, e.g., `LoadResultsFromPreviousLociStream`



_Code quote:_
```Smalltalk
await System.Linq.AsyncEnumerable.ToListAsync(previousLociResultStream)).ToDictionary(x => x.DonorId, x => x
```
___
*[`Atlas.MatchingAlgorithm/Services/Search/Matching/MatchingService.cs` line 107 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_amtJ608wIueUR0lU:-Nu_amtJ608wIueUR0lV:b-wn930w) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/Matching/MatchingService.cs#L107)):*
> ```Smalltalk
>                 searchLogger.SendTrace(string.Format(messageTemplate, count), LogLevel.Verbose);
> 
>         }
> ```

have to admit, I am not sure what `reportMask` is doing here... :sweat_smile: 

Are you saying you only want to log the message when `count` is a certain value compared to `reportMask`?



_Code quote:_
```Smalltalk
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void LogReturnedRecordsCount(int count, string messageTemplate)
        {
            const int reportMask = 0x1fff;

            if ((count & reportMask) > 0)
                searchLogger.SendTrace(string.Format(messageTemplate, count), LogLevel.Verbose);

        }
```

**Review by zabeen — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 8 unresolved discussions (waiting on @SergeyEzh)

*[a discussion](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_czuF-rcoMqLQb7lh:-Nu_czuF-rcoMqLQb7li:b-ed8ac7) (no related file):*
Now you have added all this extra logging, I wonder whether we should actually release it as a permanent change, or instead merge the commit to a non-rc branch that we just use for the matching investigations?

___

**Review by zabeen — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 8 unresolved discussions (waiting on @SergeyEzh)

*[a discussion](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_czuF-rcoMqLQb7lh:-Nu_dbzaCM8kQAEaE8VD:bw05hu5) (no related file):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

Now you have added all this extra logging, I wonder whether we should actually release it as a permanent change, or instead merge the commit to a non-rc branch that we just use for the matching investigations?
</blockquote></details>

rc/2.0.3 release to UAT-WMDA-ATLAS is blocked as I am stuck with the terraform changes to the service plan - it's not working nicely 

So, to unblock the investigation, we should deploy the logging commit via a non-rc feature branch that is still based on `stable/2.0.2` - and then we can decide if it should be released to prod at a later stage.

___

**Review by zabeen — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 8 unresolved discussions (waiting on @SergeyEzh)

*[a discussion](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_czuF-rcoMqLQb7lh:-Nudp_JQ9PYY1Dp8VMWO:b-62xaz3) (no related file):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

rc/2.0.3 release to UAT-WMDA-ATLAS is blocked as I am stuck with the terraform changes to the service plan - it's not working nicely 

So, to unblock the investigation, we should deploy the logging commit via a non-rc feature branch that is still based on `stable/2.0.2` - and then we can decide if it should be released to prod at a later stage.
</blockquote></details>

@SergeyEzh Further to above comments, rc/2.0.3 is now unblocked, but I'd rather just keep it for the terraform service plan changes. Please continue with suggestion plan to run investigation on UAT-WMDA-ATLAS via a non-rc feature branch. If you want to work on this tomorrow while I am away, please tag @daria-sorokina-da for PR approval, else we'll catch up on Monday

___

**Review by zabeen — CHANGES_REQUESTED** [review]
> Reviewed 3 of 9 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 3 unresolved discussions (waiting on @SergeyEzh)

___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 85 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_RgMFBjPxURo50xdh:-NuxiK8hAoOfEaSwutpC:bb7dx2z) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L85)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

"has completed scoring"
</blockquote></details>

bump

**Review by SergeyEzh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 2 unresolved discussions (waiting on @zabeen)

___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 85 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_RgMFBjPxURo50xdh:-NuxMvk7Bi0ko7DfZ6EJ:bpwgnnp) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L85)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

bump
</blockquote></details>

Is it accurate? By the time this message appeared in the log, this part `return rankingService.RankSearchResults(reifiedScoredResults);` of `StreamScoring` won't be actually performed? Doesn't "scoring" include sorting?
___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 93 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_TXpC6P5WKJWzZJLD:-NuxQM3J8A8uIeJDmlyn:bzao5rw) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L93)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

what do you mean by before stage 1? 

I think filtering by registry code is being done this late, because we didn't use to store registry code in the matching db until a few months ago, and had to fetch the data from the main donor db - so we did that as a one-off step to avoid repeated calls to the external db.

Now that we have that data "locally" we could shift the filtering to same place all the other filtering is done:
https://github.com/Anthony-Nolan/Atlas/blob/master/Atlas.MatchingAlgorithm/Services/Search/Matching/MatchingService.cs#L104

I'm not sure it would improve performance, but it would be cleaner.

Again, think we should do that in #1271
</blockquote></details>

I think it can improve performance if this filtering will be applied on sql server side and will be used by clients. I don't know yet much about matching algorithm, but I suppose it is more complicated than checking columns for equality to a value. The filtering can be done on indexed data very quickly, and therefore reduce dataset on which we perform matching.
___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 98 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu__Alw-r11A2U1Ta9u:-NuxRyckBV_2QriSzEfF:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L98)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

filtering results
</blockquote></details>

Done.
___
*[`Atlas.MatchingAlgorithm/Services/Search/Matching/DonorMatchingService.cs` line 114 at r1](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-NuZPPDJ7pDswYRnrpxG:-Nuxa73H8wzw0BJBfAv0:b-896fix) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/ce5bad879a712b56956b6d85637abc3259188b0d/Atlas.MatchingAlgorithm/Services/Search/Matching/DonorMatchingService.cs#L114)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

suggest tidying up by moving this line, along with the two new logging steps, to a new method, e.g., `LoadResultsFromPreviousLociStream`
</blockquote></details>

Done.
___
*[`Atlas.MatchingAlgorithm/Services/Search/Matching/MatchingService.cs` line 107 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_amtJ608wIueUR0lU:-NuxaBdJ67_YmBCK9qHl:b-gy5lba) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/Matching/MatchingService.cs#L107)):*
<details><summary><i>Previously, zabeen (Dr Zabeen Patel) wrote…</i></summary><blockquote>

have to admit, I am not sure what `reportMask` is doing here... :sweat_smile: 

Are you saying you only want to log the message when `count` is a certain value compared to `reportMask`?
</blockquote></details>

Basically, `count & 0x1fff == 0` is equal to `count % 8192 == 0` i.e. put message in the log every 8129 records. But much faster since we don't use division. Since there's a lot of records it can affect on performance. 

Thank you for the comment, I spotted mistake I made here.

**Review by zabeen — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: all files reviewed, 1 unresolved discussion (waiting on @SergeyEzh)

___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 85 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_RgMFBjPxURo50xdh:-NuxjoAK0qKMf-Nbv7Fd:bbxfgl9) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L85)):*
<details><summary><i>Previously, SergeyEzh wrote…</i></summary><blockquote>

Is it accurate? By the time this message appeared in the log, this part `return rankingService.RankSearchResults(reifiedScoredResults);` of `StreamScoring` won't be actually performed? Doesn't "scoring" include sorting?
</blockquote></details>

ranking and scoring are different -  but tbh, it doesn't matter so I am retracting!

**Review by zabeen — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @SergeyEzh)

___
*[`Atlas.MatchingAlgorithm/Services/Search/SearchService.cs` line 93 at r2](https://reviewable.io/reviews/Anthony-Nolan/Atlas/1270#-Nu_TXpC6P5WKJWzZJLD:-NuxkUXB7ET-UxB0TOOZ:b-vflmbi) ([raw file](https://github.com/Anthony-Nolan/Atlas/blob/f0c788d46f0ce022514cdb675222d82fb9dfe903/Atlas.MatchingAlgorithm/Services/Search/SearchService.cs#L93)):*
<details><summary><i>Previously, SergeyEzh wrote…</i></summary><blockquote>

I think it can improve performance if this filtering will be applied on sql server side and will be used by clients. I don't know yet much about matching algorithm, but I suppose it is more complicated than checking columns for equality to a value. The filtering can be done on indexed data very quickly, and therefore reduce dataset on which we perform matching.
</blockquote></details>

we can discuss sql level filtering - this is something I have thought about too, I think it's not straight forward because of complexity of matching sub-searches... but we should discuss soon
